principles.rst

```diff
@@ -36,55 +36,273 @@ but lacks the unarticulated principles, which will be added over time.
+that function.
+
+Once we have the `Explicit Variable Principle`_, there will never be a need for ``Proxy``.
+
+*Motivation:* As GHC supports more and more type-level programming, the ability
+to write out type signatures, arguments, and annotations has become increasingly
+important. With ``-XScopedTypeVariables``, GHC allows us to bring type variables
+into scope, but often requires us to do so by cumbersome matching. If we have
+a type ``Maybe (Either (Int, Bool, a) Double)``, that's a lot to type just to
+be able to, say, bind ``a``. The EVP says we do *not* have to resort to matching,
+ever.
+
+From `#448`_.
+
+Visibility Orthogonality Principle (VOP)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. _`Visibility Orthogonality Principle`:
 
-.. _LSP:
+**Principle**: Whether an argument is visible or
+invisible should affect only its visibility, not other properties.
 
-**Lexical Scoping Principle (LSP)**. (From `#378`_.) For every *occurrence* of an
+A consequence of the `Visibility Orthogonality Principle`_ is that these two programs should have the same meaning::
+
+   f1 :: forall a -> ...
+   f1 blah1 = ...
+
+   g1 = ... f1 blah2 ...
+
+   -------
+
+   f2 :: forall a. ...
+   f2 @(blah1) = ...
+
+   g2 = ... f2 @(blah2) ...
+
+The only difference between these is the visibility.
+
+Put another way: two programs that are the same except for visibility markers (such as
+the ``.`` vs ``->`` in a ``forall`` or the presence or absence of a ``@``) should desugar
+to the same Core program.
+
+Currently, the design for `#281`_ (along with the design for ``-XTypeApplications``)
+violates the `Visibility Orthogonality Principle`_, because the visibility marker ``@`` also affects the difference between
+term-syntax and type-syntax. However, given the `Syntactic Unification Principle`_, we strive to uphold the `Visibility Orthogonality Principle`_ when
+there is an absence of punning.
+
+*Motivation:* Visibility should be just that: a superficial property that describes
+(only) whether an argument is visible in the user-written source code.
+
+Name resolution and scoping
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Lexical Scoping Principle (LSP)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. _`Lexical Scoping Principle`:
+
+**Principle**:
+
+a. For every appearance of
+an identifier, it is possible to determine whether that appearance is a *binding site*
+or an *occurrence*, without involving the type system.
+
+b. For every *occurrence* of an
 identifier, it is possible to uniquely identify its *binding site*, without
 involving the type system.
 
-The LSP_ is true today, though Template Haskell splices may need to be run before
-completing name resolution (and running those splices requires type-checking them).
+The `Lexical Scoping Principle`_ is true today, with two complications:
+
+1. Template Haskell splices may need to be run before completing name resolution (and running those splices requires type-checking them).
```

Comment:
I can't tell, are you suggesting that TH would have to be rejected for violating LSP if it were proposed today? What is the alternative for a macro system?

If you're not suggesting that TH would be rejected, why is it ok for TH to violate LSP but not other features like a hypothetical type-directed name resolution?

(Also, do ImplicitParams violate LSP? I'm genuinely unsure.)

Reply:
Ruet and Racket being stricter about hygiene show what we should aspire to.

Reply:
I generally agree with @Ericson2314 here. I don't think TH should be rejected because of this, but TH is the primary (only?) reason that GHC actually has `type RnM = TcM` instead of properly separate monads. So it's a real drawback of TH. A better notion of staging would clarify this, because then we could run one stage entirely before starting the next one.

TDNR might be another reason to violate this principle. The point of these principles isn't so much "let's never do this again" but more to acknowledge that upholding the principles is a good thing, and thus we should have a concrete reason not to.